File: connect_4_minimax.py
import numpy as np
# import pickle


class Connect4State:

    def __init__(self, n_rows, n_cols, board=None):

        # store board variables
        self.n_rows = n_rows
        self.n_cols = n_cols

        # get board configuration from parent node if it is not the root node
        # compute available columns for the state
        self.board          = board if board is not None else np.zeros((self.n_rows, self.n_cols), dtype=int)
        self.allowed_moves  = self.init_allowed_moves()

    
    def __str__(self):

        s = ""
        # print board rows in reverse order
        # if only the columns chosen are used, then same states with distinct histories will have different string representations
        # using whole board is instead a unique identifier
        for row in self.board[::-1]:
            for col in row:
                s += str(col)
            s += "-"
        s = s[:-1]

        return s
    
    
    def init_allowed_moves(self):
        
        l = []
        for col in range(self.n_cols):

            # all columns with an empty space are allowed
            if np.any(self.board[:, col] == 0):
                l.append(col)
        l = np.array(l)
        
        # order columns by distance from the center
        # this is done to evaluate central moves first in the minimax and prune the tree faster
        sort_idx = np.argsort(np.abs(l - self.n_cols/2 + 1/self.n_cols))
        l_sort = l[sort_idx]
        
        return l_sort


    def get_win_moves(self, player):

        l = []
        for col in self.allowed_moves:

            # evaluate all possible moves
            # store in the list all columns that immediately lead to a win
            child = self.add_disc(col, player)
            if child.check_alignments(4, player):
                l.append(col)
                break
    
        return l
    

    def get_lose_moves(self, player):
        
        l = []
        for col in self.allowed_moves:

            # evaluate all possible moves
            # switch player to evaluate all possible moves for the opponent
            child   = self.add_disc(col, player)
            player_ = 3-player
            for col_ in child.allowed_moves:

                # evaluate all possible moves for the opponent
                # store in the list all columns that immediately lead to a loss if the opponent play optimally
                child_ = child.add_disc(col_, player_)
                if child_.check_alignments(4, player_):
                    l.append(col)
                    break
        
        return l
    

    def sort_moves(self, cols, player):

        alignments_count = np.zeros(len(cols))
        for c in range(len(cols)):

            # count the number of favourable alignments for each move
            # 3+ alignments are considered favourable
            child = self.add_disc(cols[c], player)
            alignments_count[c] = child.check_alignments(3, player, return_count=True)
        
        # sort columns by number of alignments
        sort_idx = np.argsort(alignments_count)[::-1]

        l_sort = np.empty(len(cols), dtype=int)
        for i in range(len(cols)):

            # if there is a draw, then keep the original ordering
            count_eq = alignments_count == alignments_count[sort_idx[i]]
            if np.sum(count_eq) > 1:

                # get columns indices with the same count
                max_count_col   = alignments_count[sort_idx[i]]
                count_eq_idx    = np.where(count_eq)

                replace_idx     = []
                for j in range(len(cols)):
                    # if a column has the same number of alignments as the maximum count, it must be replaced
                    if alignments_count[sort_idx[j]] == max_count_col:
                        replace_idx.append(j)

                # replace the column index with the static index
                # this also guarantees no repetitions
                l_sort[replace_idx] = cols[count_eq_idx]
            else:
                # otherwise, keep the sorted ordering
                l_sort[i] = cols[sort_idx[i]]
        
        return l_sort


    def get_moves(self, player):

        # if there are some winning moves, then just return them
        win_cols = self.get_win_moves(player)
        if len(win_cols) > 0:
            return win_cols

        l = self.allowed_moves.copy()
        for col in self.get_lose_moves(player):
            # return just the moves that do not lead to an immediate loss
            # remove the losing columns from the allowed columns
            l = np.delete(l, np.where(l == col)) if len(l) > 1 else l
        
        return l
    

    def add_disc(self, idx_c, player):

        # create new board and add disc to the chosen column
        board_n                 = self.board.copy()
        idx_r                   = np.argmax(self.board[:, idx_c] == 0) if np.sum(self.board[:, idx_c] > 0) else 0
        board_n[idx_r, idx_c]   = player

        # return new state
        # storing all children for all states would be too memory-intensive and with redundant information
        # this is not properly a tree, but it acts as such
        return Connect4State(
            n_rows  = self.n_rows,
            n_cols  = self.n_cols,
            board   = board_n
        )


    def score_fun(self, player):
        return n_rows*n_cols//2 + 1 - np.sum(self.board == player)


    def check_alignments(self, len, player):

        # check alignments if there are more rows than columns
        # if that is the case, then more diagonals on the width than on the height needs to be checked
        for row in range(self.n_rows):

            # check all possible horizontal alignments
            for col in range(self.n_cols-len+1):
                board_sub = self.board[row, col:col+len]
                if np.all(board_sub == player):
                    return True
            
            # check all possible primary and secondary diagonal alignments
            if row < self.n_rows-len+1:
                diag1, diag2 = get_diagonals(self.board, row)
                if np.all(diag1 == player) or np.all(diag2 == player):
                    return True
        
        # check alignments if there are more columns than rows
        # if that is the case, then more diagonals on the height than on the width needs to be checked
        for col in range(self.n_cols):

            # check all possible vertical alignments
            for row in range(self.n_rows-len+1):
                board_sub = self.board[row:row+len, col]
                if np.all(board_sub == player):
                    return True
            
            # check all possible primary and secondary diagonal alignments
            if col in range(1, self.n_cols-len+1):
                diag1, diag2 = get_diagonals(self.board, col)
                if np.all(diag1 == player) or np.all(diag2 == player):
                    return True
        
        return False
    

    def eval(self, player):

        # if the opponent has won, then a negative score is returned
        prev_player = 3-player
        if self.check_alignments(4, prev_player):
            return -self.score_fun(prev_player)
        else:
            return 0

    
    def is_full(self):
        return not np.any(self.board == 0)


    def is_root(self):
        return np.all(self.board == 0)
    

    def is_terminal(self):
        return self.check_alignments(4, 1) or self.is_full() or self.check_alignments(4, 2)


### ---------------------------------------------------------------------------------------------------- ###
### HELPER FUNCTIONS


def get_diagonals(arr, idx):

    # get main and secondary diagonal separately
    diag1 = np.diag(arr, k=idx)
    diag2 = np.diag(np.fliplr(arr), k=idx)

    return diag1, diag2


### ---------------------------------------------------------------------------------------------------- ###
### NEGAMAX ALGORITHM


def negamax(state, depth, maximizer, alpha=-float("inf"), beta=+float("inf"), memo=None):

    # convert maximizer boolean to player number
    # this is done to indicate the moves on the board and produce an identifier
    player = 1 if maximizer else 2

    if depth == 0 or state.is_terminal():
        # return evaluation for the player at terminal state
        # with negamax we are not distinguishing between maximizer and minimizer and the evaluation is the same
        # evaluate wrt the previous player
        return None, state.eval(player), memo
    
    if memo is None:
        # initialize memo at first call
        # this is also useful to reduce computational time in the search
        # acts as a hash table
        memo = dict()

    if str(state) in memo and memo[str(state)]["eval"] > 0:
        # return memoized value if optimized
        return memo[str(state)]["move"], memo[str(state)]["eval"], memo
    
    # initialize best evaluation and best move
    best_move = None
    best_eval = -np.inf

    for column in state.get_moves(player):
        # call recursion on children to evaluate all possible moves
        child           = state.add_disc(column, player)
        _, eval, memo   = negamax(child, depth-1, not maximizer, -beta, -alpha, memo)
        # switch sign of evaluation for current player
        eval            = -eval

        # update best move and evaluation
        if eval > best_eval:
            best_move = column
            best_eval = eval
        
        # prune tree if best evaluation was already found
        # this is done to reduce computation
        alpha = max(alpha, eval)
        if beta <= alpha:
            break

    # store best move and evaluation in memo
    # this is guaranteed to be the best move and evaluation for the current state
    memo[str(state)] = {
        "move": best_move,
        "eval": best_eval
    }

    return best_move, best_eval, memo


n_rows  = 5
n_cols  = 6

root    = Connect4State(n_rows, n_cols)

_, _, memo = negamax(root, 5, True)
print(memo)

# with open(f"assets/connect 4/solution_{n_rows}x{n_cols}.pkl", "wb") as f:
#     pickle.dump(memo, f)

File: connect4_minimax.c
#include <limits.h>
#include <stdbool.h>
#include <stdint.h>
#include <stdio.h>
#include <stdlib.h>


// -------------------------------------------------- //
// --- CONSTANTS AND TYPES -------------------------- //


#define ROWS 6
#define HEIGHT (ROWS+1)
#define COLS 7


typedef struct {
    uint64_t position;  // Current player chips
    uint64_t mask;      // All chips
    int n_moves;
} Board;


// -------------------------------------------------- //
// --- BITBOARD ------------------------------------- //


/**
 * Generate a bitmask representing the bottom row of the board.
 * It looks like 00...0100000010...1...
 *                     >------<>---<
 *                         6+1   7
 *               >----<>--------------<
 *                 15         49
 * Where the only 1s are evenly spaced every 7 bits.
 */
static inline uint64_t bottom_mask() {
    uint64_t mask = 0;
    for (int col = 0; col < COLS; col ++) {
        // (1ULL << (col*HEIGHT)) shifts 1 by col*HEIGHT positions to get the 1st row of each column
        // Then just take the or to set the bit
        mask |= (1ULL << (col*HEIGHT));
    }
    return mask;
}

static inline uint64_t bottom_mask_col(int col) {
    // Just take the first row of the column
    return 1ULL << (col*HEIGHT);
}

/**
 * Generate a bitmask representing all valid positions of the board.
 * It looks like 00...0011111101..101..
 *                     >-----<>---<
 *                       6+1    7
 *               >----<>--------------<
 *                 15         49
 * Where the 1s are the first 6 rows for each column.
 */
static inline uint64_t board_mask() {
    // The binary multiplication by 0b(1ULL << HEIGHT)-1 = 63d turns any 1 into 111111
    return bottom_mask() * ((1ULL << HEIGHT)-1);
}

static inline uint64_t top_mask_col(int col) {
    // Just take the sixth row of the column
    return 1ULL << (col*HEIGHT + ROWS-1);

}
 
static inline uint64_t col_mask(int col) {
    // Just take the column bits
    return ((1ULL << ROWS)-1) << (col*HEIGHT);
}


// -------------------------------------------------- //
// --- BOARD OPERATIONS ----------------------------- //


void init_board(Board *board) {
    board->position = 0;
    board->mask = 0;
    board->n_moves = 0;
}

/**
 * Check whether the current player is in a winning position.
 * This is done in 3 phases:
 *      1. Shift the board by 1 position and check that there are 2 consecutive 1-bits
 *      2. Store the shifted bool board
 *      3. Shift the board again by two positions and check that there are 3 consecutive pairs of 1-bits
 *      => If so, there is a win
 * 
 * The process is repeated for each direction.
 */
bool check_win(uint64_t pos) {
    
    // Horizontal
    uint64_t pos_shift = pos & (pos >> HEIGHT);
    if(pos_shift & (pos_shift >> (2*HEIGHT))) return true;
    
    // 1st and 2nd diagonal
    pos_shift = pos & (pos >> ROWS);
    if(pos_shift & (pos_shift >> (2*ROWS))) return true;
    pos_shift = pos & (pos >> (HEIGHT+1));
    if(pos_shift & (pos_shift >> (2*(HEIGHT+1)))) return true;
    
    // Vertical
    pos_shift = pos & (pos >> 1);
    if(pos_shift & (pos_shift >> 2)) return true;
    
    return false;
}

bool is_valid_move(Board *board, int col) {
    // Check if the top row is free
    return (board->mask & top_mask_col(col)) == 0;
}

bool is_winning_move(Board *board, int col) {
    // Check if the move yields an immediate win to the current player
    uint64_t pos = board->position;
    pos |= (board->mask + bottom_mask_col(col)) & col_mask(col);
    return check_win(pos);
}

void make_move(Board *board, int col) {
    // Switch the player
    board->position ^= board->mask;
    // Set the bottom row of the column to 1
    board->mask |= board->mask + bottom_mask_col(col);
    board->n_moves ++;
}

bool is_full(Board *board) {
    return board->n_moves == ROWS*COLS;
}


// -------------------------------------------------- //
// --- POSITION EVALUATION -------------------------- //


int evaluate_position(Board *board) {
    return 0;
}


// -------------------------------------------------- //
// --- MOVE ORDERING -------------------------------- //


// First explore moves in the center
static const int static_ordering[COLS] = {3, 2, 4, 1, 5, 0, 6};


// -------------------------------------------------- //
// --- MINIMAX ALGORITHM ---------------------------- //


int negamax(Board *board, int depth, int alpha, int beta) {
    
    // Return the evaluation at terminal state
    if (depth == 0) return evaluate_position(board);
    if (is_full(board)) return 0;

    // Check if there is an immediate win
    for (int col = 0; col < COLS; col++) {
        if (is_valid_move(board, col) && is_winning_move(board, col)) {
            return ROWS*COLS+1-board->n_moves;
        }
    }

    // Initialize
    int best_eval = INT_MIN;
    
    // Loop over the columns
    for (int idx = 0; idx < COLS; idx++) {
        int col = static_ordering[idx];
        if (is_valid_move(board, col)) {
            Board child = *board;
            make_move(&child, col);
            int eval = -negamax(&child, depth-1, -beta, -alpha);
            best_eval = (eval > best_eval) ? eval : best_eval;
            alpha = (eval > alpha) ? eval : alpha;
            if (alpha >= beta) break;
        }
    }
    return best_eval;
}

int find_best_move(Board *board, int depth) {

    // Check if there is an immediate win
    for (int col = 0; col < COLS; col++) {
        if (is_valid_move(board, col) && is_winning_move(board, col)) {
            return col;
        }
    }

    // Initialize
    int best_move = -1;
    int best_eval = INT_MIN;
    int alpha = INT_MIN;
    int beta = INT_MAX;

    // Loop over the columns
    for (int idx = 0; idx < COLS; idx++) {
        int col = static_ordering[idx];
        if (is_valid_move(board, col)) {
            Board child = *board;
            make_move(&child, col);
            int eval = -negamax(&child, depth-1, -beta, -alpha);

            if (eval > best_eval) {
                best_eval = eval;
                best_move = col;
            }
            alpha = (eval > alpha) ? eval : alpha;
            if (alpha >= beta) break;
        }
    }
    return best_move;
}


// -------------------------------------------------- //
// --- PRINT ---------------------------------------- //


void print_board(Board *board) {
    for (int row = ROWS-1; row >= 0; row--) {
        for (int col = 0; col < COLS; col++) {
            uint64_t idx = 1ULL << (col*HEIGHT + row);
            if (board->mask & idx) {
                printf("%c ", (board->position & idx) ? 'x' : 'o');
            } else {
                printf(". ");
            }
        }
        printf("\n");
    }
    printf("\n");
}


// -------------------------------------------------- //
// --- MAIN FUNCTION -------------------------------- //


int main(int argc, char *argv[]) {
    if (argc < 2) {
        printf("Usage: %s <depth>\n", argv[0]);
        return 1;
    }
    int depth = atoi(argv[1]);

    Board board;
    init_board(&board);
    int p1_move;
    int p2_move;

    while (true) {
        
        p1_move = find_best_move(&board, depth);
        make_move(&board, p1_move);
        if (check_win(board.position ^ board.mask)) {
            print_board(&board);
            printf("P1 wins!\n");
            break;
        }
        
        print_board(&board);
        printf("P2 -> Enter column (0-6): ");
        scanf("%d", &p2_move);
        if (!is_valid_move(&board, p2_move)) {
            printf("Invalid move.\n");
            continue;
        }
        make_move(&board, p2_move);

        if (check_win(board.position ^ board.mask)) {
            print_board(&board);
            printf("P2 wins!\n");
            break;
        }
    
        if (is_full(&board)) {
            print_board(&board);
            printf("It's a draw!\n");
            break;
        }
    }
    return 0;
}